### PR TITLE
next/prev: Fetch descendants with more correctness

### PR DESCRIPTION
**Pull request by essiene**
See context in [this discussion](https://github.com/martinvonz/jj/pull/3935#discussion_r1649520967)

This is the last planned change for this issue, so marking this as the fixing PR.

Fixes https://github.com/martinvonz/jj/issues/3947

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/movement_util.rs
Comment (yuja) [inline cli/src/movement_util.rs:1]: Can you also add some tests?

File: cli/src/movement_util.rs [hunk 1]
```diff
@@ -109,16 +109,20 @@ impl Direction {
...
+                        .children()
+                        .minus(working_revset)
+                        .descendants()
+                        .filtered(RevsetFilterPredicate::HasConflict)
+                        .roots()
+                } else {
+                    start_revset
+                        .descendants_at(args.offset)
+                        .minus(working_revset)
+                }
```
[the remark this answers is not in the document]
Comment (yuja) [reply cli/src/movement_util.rs:124]: BTW, I think the query can be reorganized as
```rust
let next = if !args.edit {
    start.children().minus(wc)
} else {
    start.children()
};
if args.conflict {
    next.descendants().filtered(HasConflict).roots()
} else {
    next.descendants_at(N - 1)
}
```

[the remark this answers is not in the document]
Comment (essiene) [reply cli/src/movement_util.rs]: Thanks, this was helpful. I'm finally getting used to revsets!

I've folded this in, though for consistency, I've used to match expressions so that it's clearer how `Direction::Prev` is treated. 

Using only one match expressions ends up with too much duplicated code that is just so slightly different, but a single match seems just right.

File: cli/src/movement_util.rs
Comment (essiene) [reply cli/src/movement_util.rs:1]: Done. I think I got all the interesting corner cases.

File: cli/tests/test_next_prev_commands.rs [hunk 2]
```diff
@@ -1112,6 +1112,405 @@ fn test_movement_edit_mode_false() {
     "###);
 }
 
+#[test]
+fn test_movement_target_revset() {
```
Comment (yuja) [inline cli/tests/test_next_prev_commands.rs:1116]: Some of the added tests look redundant. What we need to test in this patch is `movement.edit = false` but the wc commit has children (i.e. currently editing), right?

**Review by yuja — COMMENTED** [review]
The code change looks good to me, thanks.

File: cli/tests/test_next_prev_commands.rs (hunk 2 above)
Comment (essiene) [reply cli/tests/test_next_prev_commands.rs:1116]: Good point. I got carried away and added test that would already be covered by other tests. D'oh!

I've pruned most of them now. In particular, I've remove `edit = true` and any `prev` related stanzas.

PTAL.

Comment (yuja) [reply cli/tests/test_next_prev_commands.rs:1116]: You'll need to add tests in which `@` isn't a head. Try run the added tests against the previous commit to see if they fail.

Comment (essiene) [reply cli/tests/test_next_prev_commands.rs]: The only case I've found which isn't covered existing tests is the case where all the following hold.

* edit=false
* wc is not a head
* no conflict
* offset > 1 (I still added a test, but it's not really interesting)

I've added a test for that and confirm that it fails without the code change in this PR. See #4484 

The conflict case doesn't have this problem in trunk; when `edit=false` since we end up taking the `root()` conflicted descendants and then removing the wc.

Let me know if there's any other edge case that would make sense to explicitly test for.

File: cli/src/movement_util.rs
Comment (yuja) [inline cli/src/movement_util.rs:1]: nit: can you add `next/prev: fetch descendants ...` to the commit description header?

File: cli/tests/test_next_prev_commands.rs (hunk 2 above)
Comment (yuja) [reply cli/tests/test_next_prev_commands.rs]: Thanks. I think we could add test in which both sides have conflicts 
(you'll get prompt without this patch), but that's not so important. I think the added test is good enough.

**Review by yuja — APPROVED** [review]
Thanks!

File: cli/tests/test_next_prev_commands.rs (hunk 2 above)
Comment (essiene) [reply cli/tests/test_next_prev_commands.rs]: Thanks. Resolving now.

File: cli/src/movement_util.rs
Comment (essiene) [reply cli/src/movement_util.rs:1]: Done.
